packages/askar/src/storage/AskarStorageService.ts

```diff
@@ -0,0 +1,291 @@
+    }
+  }
+
+  /** @inheritDoc */
+  public async update(agentContext: AgentContext, record: T): Promise<void> {
+    assertAskarWallet(agentContext.wallet)
+    const session = (agentContext.wallet as AskarWallet).session
+
+    const value = JsonTransformer.serialize(record)
+    const tags = this.transformFromRecordTagValues(record.getTags()) as Record<string, string>
+
+    try {
+      await session.replace({ category: record.type, name: record.id, value, tags })
+    } catch (error) {
+      if (isAskarError(error) && error.code === askarErrors.NotFound) {
+        throw new RecordNotFoundError(`record with id ${record.id} not found.`, {
+          recordType: record.type,
+          cause: error,
+        })
+      }
+
+      throw new WalletError('Error updating record', { cause: error })
+    }
+  }
+
+  /** @inheritDoc */
+  public async delete(agentContext: AgentContext, record: T) {
+    assertAskarWallet(agentContext.wallet)
+    const session = (agentContext.wallet as AskarWallet).session
+
+    try {
+      await session.remove({ category: record.type, name: record.id })
+    } catch (error) {
+      if (isAskarError(error) && error.code === askarErrors.NotFound) {
+        throw new RecordNotFoundError(`record with id ${record.id} not found.`, {
+          recordType: record.type,
+          cause: error,
+        })
+      }
+      throw new WalletError('Error deleting record', { cause: error })
+    }
+  }
+
+  /** @inheritDoc */
+  public async deleteById(
+    agentContext: AgentContext,
+    recordClass: BaseRecordConstructor<T>,
+    id: string
+  ): Promise<void> {
+    assertAskarWallet(agentContext.wallet)
+    const session = (agentContext.wallet as AskarWallet).session
+
+    try {
+      await session.remove({ category: recordClass.type, name: id })
+    } catch (error) {
+      if (isAskarError(error) && error.code === askarErrors.NotFound) {
+        throw new RecordNotFoundError(`record with id ${id} not found.`, {
+          recordType: recordClass.type,
+          cause: error,
+        })
+      }
+      throw new WalletError('Error deleting record', { cause: error })
+    }
+  }
+
+  /** @inheritDoc */
+  public async getById(agentContext: AgentContext, recordClass: BaseRecordConstructor<T>, id: string): Promise<T> {
+    assertAskarWallet(agentContext.wallet)
+    const session = (agentContext.wallet as AskarWallet).session
+
+    try {
+      const record = await session.fetch({ category: recordClass.type, name: id })
+      if (!record) {
+        throw new RecordNotFoundError(`record with id ${id} not found.`, {
+          recordType: recordClass.type,
+        })
+      }
+      return this.recordToInstance(record, recordClass)
+    } catch (error) {
+      if (
+        isAskarError(error) &&
+        (error.code === askarErrors.NotFound ||
+          // FIXME: this is current output from askar wrapper but does not describe specifically a not found scenario
```

Comment:
It seems Askar also just returns not found if the record wasn't found: 

```
        try:
            item = await self._session.handle.fetch(
                record_type, record_id, for_update=for_update
            )
        except AskarError as err:
            raise StorageError("Error when fetching storage record") from err
        if not item:
            raise StorageNotFoundError(f"Record not found: {record_type}/{record_id}")
```

Will it sometimes throw an error instead of not returning a record?

Reply:
Not 100% sure but I'd say that it always sends the error.code Custom (100). Let's analyze on Askar side to see where the problem is.